File: src/cache.ts
/**
 * Local cache layout (spec §9):
 *
 *   objects/sha256/<64hex>.json    canonical signed objects/manifests
 *   evidence/sha256/<64hex>.bin    explicitly imported raw evidence
 *   bundles/<64hex>.json           exported bundles
 *   pending/<slq_id>.json          durable commands awaiting a receipt
 *   pins/<sll_id>.json             largest verified SignedHead per ledger
 *   journal.ndjson                 local.* lifecycle records
 *
 * Directories are 0700, files 0600; temp files are fsynced and atomically
 * renamed; content-hash filenames are derived internally.
 */

import { mkdirSync, appendFileSync, existsSync, readdirSync, unlinkSync, chmodSync } from "node:fs";
import { join } from "node:path";
import { SunlightError } from "./errors.js";
import { jcsBytes } from "./encoding/jcs.js";
import { parseStrictJson } from "./encoding/strict-json.js";
import { RAW } from "./crypto/domains.js";
import { parseSignedHead, type LocalJournalRecord, type SignedHead } from "./schema.js";
import { readFileBytesSync, writeFileAtomic } from "./fsutil.js";

export class Cache {
  readonly dir: string;
  constructor(dir: string) {
    this.dir = dir;
  }

  private ensure(): void {
    for (const sub of ["objects/sha256", "evidence/sha256", "bundles", "pending", "pins"]) {
      mkdirSync(join(this.dir, sub), { recursive: true, mode: 0o700 });
    }
    try {
      // Tighten in case umask was permissive.
      chmodSync(this.dir, 0o700);
    } catch { /* best effort */ }
  }

  private hexOf(digest: string): string {
    return RAW(digest).toString("hex");
  }

  storeObject(digest: string, obj: unknown): string {
    this.ensure();
    const p = join(this.dir, "objects/sha256", `${this.hexOf(digest)}.json`);
    if (!existsSync(p)) writeFileAtomic(p, jcsBytes(obj));
    return p;
  }

  loadObject(digest: string): unknown | null {
    const p = join(this.dir, "objects/sha256", `${this.hexOf(digest)}.json`);
    if (!existsSync(p)) return null;
    return parseStrictJson(readFileBytesSync(p));
  }

  storeEvidence(digest: string, raw: Buffer): string {
    this.ensure();
    const p = join(this.dir, "evidence/sha256", `${this.hexOf(digest)}.bin`);
    if (!existsSync(p)) writeFileAtomic(p, raw);
    return p;
  }

  storeBundle(digest: string, bundle: unknown): string {
    this.ensure();
    const p = join(this.dir, "bundles", `${this.hexOf(digest)}.json`);
    if (!existsSync(p)) writeFileAtomic(p, jcsBytes(bundle));
    return p;
  }

  /** Persist a command durably before its first network attempt. */
  storePending(commandId: string, command: unknown): string {
    this.ensure();
    if (!/^slq_[A-Za-z0-9_-]{21}$/.test(commandId)) throw new SunlightError("SCHEMA_INVALID");
    const p = join(this.dir, "pending", `${commandId}.json`);
    if (!existsSync(p)) writeFileAtomic(p, jcsBytes(command));
    return p;
  }

  dropPending(commandId: string): void {
    const p = join(this.dir, "pending", `${commandId}.json`);
    if (existsSync(p)) unlinkSync(p);
  }

  listPending(): string[] {
    const d = join(this.dir, "pending");
    if (!existsSync(d)) return [];
    return readdirSync(d)
      .filter((f) => f.endsWith(".json"))
      .map((f) => f.slice(0, -5));
  }

  loadPending(commandId: string): unknown | null {
    const p = join(this.dir, "pending", `${commandId}.json`);
    if (!existsSync(p)) return null;
    return parseStrictJson(readFileBytesSync(p));
  }

  /** Largest verified head pin per ledger; conflicts quarantine (caller). */
  loadPin(ledger: string): SignedHead | null {
    const p = join(this.dir, "pins", `${ledger}.json`);
    if (!existsSync(p)) return null;
    return parseSignedHead(parseStrictJson(readFileBytesSync(p)));
  }

  storePin(ledger: string, head: SignedHead): void {
    this.ensure();
    if (!/^sll_[A-Za-z0-9_-]{21}$/.test(ledger)) throw new SunlightError("SCHEMA_INVALID");
    writeFileAtomic(join(this.dir, "pins", `${ledger}.json`), jcsBytes(head));
  }

  journal(event: string, command: string | null, statement: string | null, atMs: number): void {
    this.ensure();
    const rec: LocalJournalRecord = {
      v: "sunlight.local/1",
      command,
      statement,
      event,
      at_ms: atMs,
    };
    appendFileSync(join(this.dir, "journal.ndjson"), jcsBytes(rec).toString("utf8") + "\n", { mode: 0o600 });
  }
}

File: python/sunlight/cache.py
"""Local cache layout (spec §9):

    objects/sha256/<64hex>.json    canonical signed objects/manifests
    evidence/sha256/<64hex>.bin    explicitly imported raw evidence
    bundles/<64hex>.json           exported bundles
    pending/<slq_id>.json          durable commands awaiting a receipt
    pins/<sll_id>.json             largest verified SignedHead per ledger
    journal.ndjson                 local.* lifecycle records

Directories are 0700, files 0600; content-hash filenames are derived
internally.
"""

from __future__ import annotations

import os
import re

from .errors import SunlightError
from .fsutil import read_file_bytes, write_file_atomic
from .jcs import jcs
from .schema import parse_signed_head
from .strictjson import parse_strict_json

_PENDING_RE = re.compile(r"^slq_[A-Za-z0-9_-]{21}$")
_LEDGER_RE = re.compile(r"^sll_[A-Za-z0-9_-]{21}$")


class Cache:
    def __init__(self, dir: str):
        self.dir = dir

    def _ensure(self) -> None:
        for sub in ("objects/sha256", "evidence/sha256", "bundles", "pending", "pins"):
            os.makedirs(os.path.join(self.dir, sub), mode=0o700, exist_ok=True)
        try:
            os.chmod(self.dir, 0o700)
        except OSError:
            pass

    @staticmethod
    def _hex(digest: str) -> str:
        if not isinstance(digest, str) or not re.match(r"^sha256:[0-9a-f]{64}$", digest):
            raise SunlightError("SCHEMA_INVALID")
        return digest[7:]

    def store_object(self, digest: str, obj) -> str:
        self._ensure()
        p = os.path.join(self.dir, "objects/sha256", f"{self._hex(digest)}.json")
        if not os.path.exists(p):
            write_file_atomic(p, jcs(obj))
        return p

    def load_object(self, digest: str):
        p = os.path.join(self.dir, "objects/sha256", f"{self._hex(digest)}.json")
        if not os.path.exists(p):
            return None
        return parse_strict_json(read_file_bytes(p))

    def store_evidence(self, digest: str, raw: bytes) -> str:
        self._ensure()
        p = os.path.join(self.dir, "evidence/sha256", f"{self._hex(digest)}.bin")
        if not os.path.exists(p):
            write_file_atomic(p, raw)
        return p

    def store_bundle(self, digest: str, bundle) -> str:
        self._ensure()
        p = os.path.join(self.dir, "bundles", f"{self._hex(digest)}.json")
        if not os.path.exists(p):
            write_file_atomic(p, jcs(bundle))
        return p

    def store_pending(self, command_id: str, command) -> str:
        self._ensure()
        if not _PENDING_RE.match(command_id):
            raise SunlightError("SCHEMA_INVALID")
        p = os.path.join(self.dir, "pending", f"{command_id}.json")
        if not os.path.exists(p):
            write_file_atomic(p, jcs(command))
        return p

    def drop_pending(self, command_id: str) -> None:
        p = os.path.join(self.dir, "pending", f"{command_id}.json")
        if os.path.exists(p):
            os.unlink(p)

    def list_pending(self) -> list[str]:
        d = os.path.join(self.dir, "pending")
        if not os.path.isdir(d):
            return []
        return [f[:-5] for f in os.listdir(d) if f.endswith(".json")]

    def load_pending(self, command_id: str):
        p = os.path.join(self.dir, "pending", f"{command_id}.json")
        if not os.path.exists(p):
            return None
        return parse_strict_json(read_file_bytes(p))

    def load_pin(self, ledger: str):
        p = os.path.join(self.dir, "pins", f"{ledger}.json")
        if not os.path.exists(p):
            return None
        return parse_signed_head(parse_strict_json(read_file_bytes(p)))

    def store_pin(self, ledger: str, head) -> None:
        self._ensure()
        if not _LEDGER_RE.match(ledger):
            raise SunlightError("SCHEMA_INVALID")
        write_file_atomic(os.path.join(self.dir, "pins", f"{ledger}.json"), jcs(head))

    def journal(self, event: str, command: str | None, statement: str | None, at_ms: int) -> None:
        self._ensure()
        rec = {
            "v": "sunlight.local/1",
            "command": command,
            "statement": statement,
            "event": event,
            "at_ms": at_ms,
        }
        fd = os.open(os.path.join(self.dir, "journal.ndjson"), os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o600)
        try:
            os.write(fd, jcs(rec) + b"\n")
        finally:
            os.close(fd)
